### PR TITLE
Scripts for publishing + doc deployment + JSDoc support

### PR DESCRIPTION
**Pull request by gauntface**
cc @jeffposnick

This is the very first step in hopefully fixing some of the problems in #3 by adding docs.

@addyosmani @wibblymat @surma anyone available for a code review?

### CODE REVIEW

File: project/publish-docs.sh [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+#         - GH_REF: github.com/<username>/<repo>.git
+#
+#########################################################################
+
+if [ "$BASH_VERSION" = '' ]; then
+ echo "    Please run this script via this command: './project/publish-docs.sh'"
+ exit 1;
+fi
+
+GITHUB_REPO="https://github.com/GoogleChrome/sw-testing-helpers/"
```
Comment (surma) [inline project/publish-docs.sh]: Should we maybe pull in this URL from either `package.json` or get it from `git`?

File: project/publish-docs.sh [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+echo "Deploying new docs"
+echo ""
+
+echo ""
+echo ""
+echo "Clone repo and get gh-pages branch"
+echo ""
+git clone $GITHUB_REPO ./gh-pages
+cd ./gh-pages
+git checkout -b gh-pages
```
Comment (surma) [inline project/publish-docs.sh]: I’d prefer `git checkout --orphan gh-pages && git clean -fxd`. The folder will be empty and it will be a stand-alone commit in the end.

File: project/publish-docs.sh (hunk 1 above)
Comment (gauntface) [inline project/publish-docs.sh]: I'd love to get it from git because it'll make the script reusable. At the moment I'm fighting an issue with travis :(

File: project/publish-docs.sh [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+  git add ./$docLocation
+  git commit -m "Deploy to GitHub Pages"
+
+  if [ "$TRAVIS" ]; then
+    # Force push from the current repo's master branch to the remote
+    # repo's gh-pages branch. (All previous history on the gh-pages branch
+    # will be lost, since we are overwriting it.) We redirect any output to
+    # /dev/null to hide any sensitive credential data that might otherwise be exposed.
+    git push --force --quiet "https://${GH_TOKEN}@${GH_REF}" gh-pages > /dev/null 2>&1
```
Comment (surma) [inline project/publish-docs.sh]: Not entirely sold of swallowing stderr here.

File: project/publish-release.sh [hunk 4]
```diff
@@ -77,9 +106,8 @@ echo ""
 echo ""
 echo "Git push to release branch"
 echo ""
-GIT_REPO="https://github.com/GoogleChrome/sw-testing-helpers.git"
 git init
-git remote add origin $GIT_REPO
+git remote add origin $GITHUB_REPO
 git checkout -b release
```
Comment (surma) [inline project/publish-release.sh]: Potentially the same as above

File: project/publish-release.sh [hunk 5]
```diff
@@ -106,6 +134,21 @@ echo ""
 
 git add package.json
 git commit -m "Auto-generated PR to update package.json with new version - $PACKAGE_VERSION"
-git push -f origin release-pr
+git push origin -f $currentBranch:release-pr
+
+# Make sure any pending commits on master are reset (i.e. the package.json commit)
+git reset --hard origin/$currentBranch
```
Comment (surma) [inline project/publish-release.sh]: also `git clean -fxd`?

File: project/publish-docs.sh (hunk 3 above)
Comment (gauntface) [inline project/publish-docs.sh]: The reason (despite being painful for debugging) is that it uses a token that will have commit access to the public repo, which is printed in the log, is publicly available on travis - hence the killing of output.

File: project/publish-docs.sh (hunk 2 above)
Comment (gauntface) [inline project/publish-docs.sh]: will have a play with these later, what is --orphan do?

Comment (surma) [inline project/publish-docs.sh]: Creates a branch that doesn’t point to any commit. If you commit something on that branch, that commit will have no parent. It’s free floating (like the first commit in a repo).

File: project/publish-docs.sh (hunk 3 above)
Comment (surma) [inline project/publish-docs.sh]: Gotcha

File: project/publish-docs.sh (hunk 2 above)
Comment (gauntface) [inline project/publish-docs.sh]: That's not what I want in that case.

Basically each release would commit a set of versioned docs to gh-pages i.e. we'd have a master and releases section and within releases we'd have v0.1.0, v0.2.0 ....etc

So we wouldn't want an orphaned branch. We just want to overwrite / add to the existing doc set of a specific directory.

File: project/publish-release.sh (hunk 5 above)
Comment (gauntface) [inline project/publish-release.sh]: Looking through this S/O post: http://stackoverflow.com/questions/1090309/git-undo-all-working-dir-changes-including-new-files

Sounds like is git reset --hard works (which so far it does seem to), I'm inclined to say leave as is. I'd rather leave behind a little mess and then have to figure out the right course of action then go for a nuclear option at the cost of losing some work - am I being too cautious?

File: project/publish-docs.sh (hunk 1 above)
Comment (gauntface) [inline project/publish-docs.sh]: This seems to work: GITHUB_REPO=$(git config --get remote.origin.url)

File: project/publish-release.sh (hunk 4 above)
Comment (gauntface) [inline project/publish-release.sh]: If this is regarding grabbing from git or package.json - switched to get from git :)

File: project/publish-release.sh (hunk 5 above)
Comment (surma) [inline project/publish-release.sh]: Nah that’s fine.

File: project/publish-docs.sh (hunk 1 above)
Comment (surma) [inline project/publish-docs.sh]: Cool. Just keep in mind that the remote URL *might* be an SSH URL, or even something weird like `gh:GoogleChrome/sw-testing-helpers` (that’s what mine would look like).
